runtime/js-compute-runtime/builtins/cache-simple.cpp

```diff
@@ -0,0 +1,440 @@
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+
+  auto body = HttpBody(body_handle);
+  if (!body.valid()) {
+    return false;
+  }
+  // source_body will only be valid when the body parameter is a Host-backed ReadableStream
+  if (source_body.valid()) {
+    auto res = body.append(source_body);
+    if (auto *err = res.to_err()) {
+      HANDLE_ERROR(cx, *err);
+      return false;
+    }
+    args.rval().setUndefined();
+    return true;
+  } else {
+    auto write_res = body.write_all(reinterpret_cast<uint8_t *>(buf.get()), length);
+    if (auto *err = write_res.to_err()) {
+      HANDLE_ERROR(cx, *err);
+      return false;
+    }
+  }
+  auto close_res = body.close();
+  if (auto *err = close_res.to_err()) {
+    HANDLE_ERROR(cx, *err);
+    return false;
+  }
+
+  args.rval().setUndefined();
+  return true;
+}
+
+// static get(key: string): SimpleCacheEntry | null;
+bool SimpleCache::get(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The SimpleCache builtin");
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "SimpleCache.get", 1)) {
+    return false;
+  }
+
+  fastly_world_string_t key;
+  // Convert key parameter into a string and check the value adheres to our validation rules.
+  JS::UniqueChars key_chars = encode(cx, args[0], &key.len);
+  if (!key_chars) {
+    return false;
+  }
+  key.ptr = key_chars.get();
+
+  if (key.len == 0) {
+    JS_ReportErrorASCII(cx, "SimpleCache.get: key can not be an empty string");
+    return false;
+  }
+  if (key.len > 8135) {
+    JS_ReportErrorASCII(cx,
+                        "SimpleCache.get: key is too long, the maximum allowed length is 8135.");
+    return false;
+  }
+
+  fastly_error_t err;
+  fastly_cache_lookup_options_t options;
+  std::memset(&options, 0, sizeof(options));
+
+  fastly_cache_handle_t handle = INVALID_HANDLE;
+  if (!fastly_cache_lookup(&key, &options, &handle, &err)) {
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+
+  fastly_body_handle_t body = INVALID_HANDLE;
+  fastly_cache_get_body_options_t opts;
+  if (!fastly_cache_get_body(handle, &opts, &body, &err)) {
+    HANDLE_ERROR(cx, err);
+    return false;
+  }
+
+  if (body == INVALID_HANDLE) {
+    args.rval().setNull();
+  } else {
+    JS::RootedObject entry(cx, SimpleCacheEntry::create(cx, body));
+    if (!entry) {
+      return false;
+    }
+    args.rval().setObject(*entry);
+  }
+
+  return true;
+}
+
+// static delete(key: string): undefined;
+bool SimpleCache::delete_(JSContext *cx, unsigned argc, JS::Value *vp) {
+  REQUEST_HANDLER_ONLY("The SimpleCache builtin");
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "SimpleCache.delete", 1)) {
+    return false;
+  }
+
+  fastly_world_string_t key;
+  // Convert key parameter into a string and check the value adheres to our validation rules.
+  {
+    JS::UniqueChars key_chars = encode(cx, args.get(0), &key.len);
```

Comment:
This needs to be moved outside of the curly braces, as it will be freed on line 388 otherwise. As a result, `key` will point at freed memory when it's used on line 393.